### PR TITLE
fix: augment universe_domain handling

### PR DESCRIPTION
**Pull request by shollyman**
This PR revisits the universe resolution for the BQ client, and handles
new requirements like env-based specification and validation.

### CODE REVIEW

File: google/cloud/bigquery/_helpers.py [hunk 1]
```diff
@@ -55,9 +57,63 @@
...
+        client_options: specified client options.
+    Returns:
+        str: resolved universe setting.
+
+    """
+    if isinstance(client_options, dict):
+        client_options = client_options_lib.from_dict(client_options)
+    universe = _DEFAULT_UNIVERSE
+    if hasattr(client_options, "universe_domain"):
+        options_universe = getattr(client_options, "universe_domain")
```
Comment (Linchin) [inline google/cloud/bigquery/_helpers.py:85]: Do we need to account for the case when `options_universe == ""`?

File: google/cloud/bigquery/client.py [hunk 2]
```diff
@@ -253,16 +257,18 @@ def __init__(
...
+                if client_universe != _DEFAULT_UNIVERSE:
+                    kw_args["api_endpoint"] = _DEFAULT_HOST_TEMPLATE.replace(
+                        "{UNIVERSE_DOMAIN}", client_universe
+                    )
+        # Ensure credentials and universe are not in conflict.
+        if hasattr(self, "_credentials") and client_universe is not None:
+            _validate_universe(client_universe, self._credentials)
 
         self._connection = Connection(self, **kw_args)
+
```
Comment (Linchin) [inline google/cloud/bigquery/client.py:271]: nit: should we keep the blank line here?

File: tests/unit/helpers.py [hunk 3]
```diff
@@ -43,6 +43,20 @@ def make_client(project="PROJECT", **kw):
     return google.cloud.bigquery.client.Client(project, credentials, **kw)
 
 
+def make_creds(creds_universe: None):
+    from google.auth import credentials
+
+    class TestingCreds(credentials.Credentials):
+        def refresh(self, request):  # pragma: NO COVER
```
Comment (Linchin) [inline tests/unit/helpers.py:50]: For my education, what is the purpose of the `refresh()` method here?

**Comment by Linchin** [comment]
Thank you @shollyman! Overall it looks good, despite a new nitpicks.

File: tests/unit/helpers.py (hunk 3 above)
Comment (shollyman) [reply tests/unit/helpers.py:50]: This subclasses credentials.  Refresh is part of the mechanism to provide bearer credentials, but is marked abstract in the base class so we need to provide an impl even though we're not using it here.

File: google/cloud/bigquery/client.py (hunk 2 above)
Comment (shollyman) [reply google/cloud/bigquery/client.py]: removed

File: google/cloud/bigquery/_helpers.py (hunk 1 above)
Comment (shollyman) [reply google/cloud/bigquery/_helpers.py:85]: good point, added conditionals and tests.

File: tests/unit/test__helpers.py [hunk 4]
```diff
@@ -17,8 +17,74 @@
...
+
+        self.assertEqual("googleapis.com", _get_client_universe(None))
+
+    def test_with_dict(self):
+        from google.cloud.bigquery._helpers import _get_client_universe
+
+        options = {"universe_domain": "foo.com"}
+        self.assertEqual("foo.com", _get_client_universe(options))
+
+    def test_with_clientoptions(self):
```
[the remark this answers is not in the document]
Comment (shollyman) [reply tests/unit/test__helpers.py]: done

**Review by Linchin — APPROVED** [review]
LGTM. Thank you!

**Pull request merged**